src/t_zset.c

```diff
@@ -3819,71 +3845,108 @@ void genericZpopCommand(client *c, robj **keyv, int keyc, int where, int emitkey
+    }
+
+    genericZpopCommand(c, &c->argv[1], 1, where, 0,
+                       count, 0, NULL);
+}
+
+/* ZPOPMIN key [<count>] */
+void zpopminCommand(client *c) {
+    zpopMinMaxCommand(c, ZSET_MIN);
 }
 
 /* ZMAXPOP key [<count>] */
 void zpopmaxCommand(client *c) {
-    if (c->argc > 3) {
-        addReplyErrorObject(c,shared.syntaxerr);
-        return;
-    }
-    genericZpopCommand(c,&c->argv[1],1,ZSET_MAX,0,
-        c->argc == 3 ? c->argv[2] : NULL);
+    zpopMinMaxCommand(c, ZSET_MAX);
 }
 
-/* BZPOPMIN / BZPOPMAX actual implementation. */
-void blockingGenericZpopCommand(client *c, int where) {
+/* BZPOPMIN, BZPOPMAX, BZMPOP actual implementation.
+ *
+ * 'numkeys' is the number of keys.
+ * 'timeout_idx' parameter position of block timeout.
+ * 'where' ZSET_MIN or ZSET_MAX.
+ * 'count' is the number of elements requested to pop, or 0 for plain single pop.
+ * 'use_nested_array_with_key' todo need a comment and a new name */
+void blockingGenericZpopCommand(client *c, robj **keys, int numkeys, int where,
+                                int timeout_idx, long count, int use_nested_array_with_key) {
     robj *o;
+    robj *key;
     mstime_t timeout;
     int j;
 
-    if (getTimeoutFromObjectOrReply(c,c->argv[c->argc-1],&timeout,UNIT_SECONDS)
+    if (getTimeoutFromObjectOrReply(c,c->argv[timeout_idx],&timeout,UNIT_SECONDS)
         != C_OK) return;
 
-    for (j = 1; j < c->argc-1; j++) {
-        o = lookupKeyWrite(c->db,c->argv[j]);
+    for (j = 0; j < numkeys; j++) {
+        key = keys[j];
+        o = lookupKeyWrite(c->db,key);
+        /* Non-existing key, move to next key. */
+        if (o == NULL) continue;
+
         if (checkType(c,o,OBJ_ZSET)) return;
-        if (o != NULL) {
-            if (zsetLength(o) != 0) {
-                /* Non empty zset, this is like a normal ZPOP[MIN|MAX]. */
-                genericZpopCommand(c,&c->argv[j],1,where,1,NULL);
-                /* Replicate it as an ZPOP[MIN|MAX] instead of BZPOP[MIN|MAX]. */
-                rewriteClientCommandVector(c,2,
-                    where == ZSET_MAX ? shared.zpopmax : shared.zpopmin,
-                    c->argv[j]);
-                return;
-            }
+
+        long llen = zsetLength(o);
+        /* Empty zset, move to next key. */
+        if (llen == 0) continue;
+
+        if (count != 0) {
+            /* BZMPOP, non empty zset, like a normal ZPOP[MIN|MAX] with count option.
+             * The difference here we pop a range of elements in a nested arrays way. */
+            genericZpopCommand(c, &key, 1, where, 1, count,
+                               use_nested_array_with_key, NULL);
+
+            /* Replicate it as ZPOP[MIN|MAX] COUNT. */
+            robj *count_obj = createStringObjectFromLongLong((count > llen) ? llen : count);
+            rewriteClientCommandVector(c, 3,
+                                       (where == ZSET_MAX) ? shared.zpopmax : shared.zpopmin,
+                                       key, count_obj);
```

Comment:
let's make sure we have tests for the propagation of this (both when it gets blocked and when it doesn't)

Reply:
I will add these tests later